[WIP] Offset manager tests

aaronkavlie-wf opened the pull request:
This PR adds tests for the offset manager in PR #461.
The majority of the code now has coverage.

Notes
--------

- I was unable to reach the `bom.updateSubscriptions` loop in the `brokerOffsetManager.abort()` method; pointers for that would be appreciated.

- The tests use a pretty simple setup; adding a second partition offset manager (in an attempt to reach the code in the previous point) led to a a broker unreference/refresh and a need for more mock responses that I didn't get to.

- After spending a while with it, the code is still somewhat difficult to reason about, due to a lot of channels and resultant side effects. Also, 3 different objects whose responsibilities are not totally clear.

@eapache @wvanbergen @horkhe

aaronkavlie-wf commented:
Latest commit includes a small bug fix in `offset_manager.go`

aaronkavlie-wf commented:
@eapache this is worth a look now; most of the code is covered. Feedback appreciated.

wvanbergen commented:
You forgot to handle some errors:

```
1.59s$ make errcheck
errcheck github.com/Shopify/sarama/...
github.com/Shopify/sarama/offset_manager_test.go:81:18	testClient.Close()
github.com/Shopify/sarama/offset_manager_test.go:130:11	pom.Close()
github.com/Shopify/sarama/offset_manager_test.go:131:18	testClient.Close()
github.com/Shopify/sarama/offset_manager_test.go:166:11	pom.Close()
github.com/Shopify/sarama/offset_manager_test.go:167:18	testClient.Close()
github.com/Shopify/sarama/offset_manager_test.go:180:11	pom.Close()
github.com/Shopify/sarama/offset_manager_test.go:183:18	testClient.Close()
github.com/Shopify/sarama/offset_manager_test.go:203:11	pom.Close()
github.com/Shopify/sarama/offset_manager_test.go:204:18	testClient.Close()
github.com/Shopify/sarama/offset_manager_test.go:277:18	testClient.Close()
github.com/Shopify/sarama/offset_manager_test.go:302:11	pom.Close()
github.com/Shopify/sarama/offset_manager_test.go:304:18	testClient.Close()
make: *** [errcheck] Error 1
```

We use errcheck to enforce that every error return value is handled (or at least explicitly ignored)

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+	config              *Config
+)
+
+func initOM(t *testing.T) OffsetManager {
```
I'd prefer to not use abbreviations for function names, so `initOffsetManager`

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+	return om
+}
+
+func initPOM(t *testing.T) PartitionOffsetManager {
```
`initPartitionOffsetManager`

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+	broker, coordinator *mockBroker
+	testClient          Client
+	config              *Config
+)
```
@eapache Do we usually have shared state like this for tests? Or do we always keep all state contained inside the test function? Not sure what to think about this.

eapache replied:
I would prefer to avoid globals on principle. the init functions can just return the additional values

aaronkavlie-wf replied:
The only reason I can remember that I went this route was to simplify return values for those init functions. Shouldn't be too difficult to fix...

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+	testClient.Close()
+
+	// Wait to see what the brokerOffsetManager does
+	time.Sleep(500 * time.Millisecond)
```
What are we waiting on here?

aaronkavlie-wf replied:
Thanks for pointing this out, it was for debugging of an infinite loop as I hadn't closed the pom.

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+	if err != nil {
+		t.Error(err)
+	}
+
```
Maybe we should check whether `pom.Offset()` has the expected value here.

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+	}
+	if meta != "modified_meta" {
+		t.Errorf("Expected metadata \"modified_meta\". Actual: %q", meta)
+	}
```
Do we need to test whether we end up doing a CommitOffset requests at some point?

wvanbergen replied:
Never mind; this is tested in the next test.

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+}
+
+// Test of recovery from abort
+func TestBOMAbort(t *testing.T) {
```
`TestAbortPartitionOffsetManager`

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+	testClient.Close()
+}
+
+func TestOffset(t *testing.T) {
```
`TestPartitionOffsetManagerOffset`

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+	testClient.Close()
+}
+
+func TestSetOffset(t *testing.T) {
```
`TestPartitionOffsetManagerSetOffset`

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,305 @@
+
+// Test recovery from ErrNotCoordinatorForConsumer
+// on first fetchInitialOffset call
+func TestFetchInitialFail(t *testing.T) {
```
`TestOffsetManagerOffsetInitialFetchFail`

wvanbergen commented:
A couple of questions and naming nitpicks, but this looks good. Thanks for helping on with this!

aaronkavlie-wf commented:
The latest batch of commits addresses all the naming comments and removes the `time.Sleep()` call.

wvanbergen commented:
Green CI :green_heart:. I am happy once the globals are done.

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,302 @@
+	coordinator.Close()
+}
+
+func TestCommitErr(t *testing.T) {
```
Forgot this one: `TestPartitionOffsetManagerCommitError`

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,302 @@
+}
+
+// Test fetchInitialOffset retry on ErrOffsetsLoadInProgress
+func TestFetchInitialInProgress(t *testing.T) {
```
`TestOffsetManagerFetchInitialLoadInProgress`

wvanbergen commented on offset_manager_test.go:
```diff
@@ -0,0 +1,302 @@
+		Err:      ErrNotCoordinatorForConsumer,
+		Offset:   5,
+		Metadata: "test_meta",
+	}
```
Maybe we should verify that the Offset and Metadata values are ignored for an error response, i.e.  `Offset()` will not return 5 after this.

aaronkavlie-wf replied:
In the test code? Sarama retries automatically on error so it appears there's no way to call a `pom.Offset()` if you get an error.

wvanbergen replied:
ah true - didn't think of that.

wvanbergen commented:
The reason I want the long test function names is that it allows you to do `go test -run OffsetManagaer` to run all tests related to the offset manager.

aaronkavlie-wf commented:
Pushed up another commit to get rid of globals.

4 return values isn't pretty; I guess the only way out of that, though, would be to define a struct or two.

wvanbergen commented:
You can always create a `testState` struct or something along those lines if you don't want to return 4 separate values.

aaronkavlie-wf commented:
> The reason I want the long test function names is that it allows you to do go test -run OffsetManager to run all tests related to the offset manager.

Yeah, I wanted to do that during testing. I wish you could target a specific file without worrying about test names.

wvanbergen commented:
I think I am happy with this now :+1: @eapache final comments?

eapache commented on offset_manager_test.go:
```diff
@@ -0,0 +1,300 @@
+	}
+
+	broker.Returns(&ConsumerMetadataResponse{
+		CoordinatorID:   coordinator.BrokerID(),
```
you should just be able to set `Coordinator: coordinator` and it will pick up the ID/Host/Port automatically

aaronkavlie-wf replied:
Tried that, I get this:

```
cannot use coordinator (type *mockBroker) as type *Broker in field value
```

nor does type assertion (`coordinator.(*Broker)`) fix it.

eapache replied:
oh right, mockBroker != Broker... nevermind then

eapache commented on offset_manager_test.go:
```diff
@@ -0,0 +1,300 @@
+	return om, testClient, broker, coordinator
+}
+
+func initPartitionOffsetManager(t *testing.T) (pom PartitionOffsetManager,
```
I'd rather this function take the offset manager as an argument (so callers would have to call `initOffsetManager` themselves first) so that all methods have access to the root manager. This will make it much easier if/when we add manager-level methods like a global `Close`

eapache commented on offset_manager_test.go:
```diff
@@ -0,0 +1,300 @@
+
+	// For RefreshCoordinator()
+	broker.Returns(&ConsumerMetadataResponse{
+		CoordinatorID:   newCoordinator.BrokerID(),
```
for all of these `ConsumerMetadataResponse`s you should just be able to set the `Coordinator` field and nothing else

eapache commented:
A few minor points, but generally :+1:. Please squash away all the intermediate commits, and rebase on the branch to make sure my recent changes haven't introduced any bugs :)

aaronkavlie-wf commented:
@eapache 

> Please squash away all the intermediate commits

squash everything here to a single commit?

eapache commented:
> squash everything here to a single commit?

It doesn't have to be a single commit, but 21 is way too many. However many makes sense for each commit to be its own self-contained and consistent step.

aaronkavlie-wf commented:
I rebased and squashed some of my commits.

I mistakenly rebased on master, then rebased again on the offset-manager branch. Still includes master stuff (rebasing the offset-manager branch ought to clean that out).

It's down to 9 commits from this branch, is that reasonable?

There is a CI failure; I'm not able to reproduce locally.

wvanbergen commented:
CI seems to be a flaky test; I restarted it.

eapache commented:
I've rebased the offset-manager branch, you should be able to rebase again to get rid of all the extra commits.

aaronkavlie-wf commented:
Rebased; it still includes a bunch of your commits in this PR for some reason.

eapache commented:
Ya, something was weird. I ended up manually grabbing both branches locally, doing all the rebase/resolution here, and then pushing the resulting commits up to the existing `offset-manager` branch.

So this has, in effect, been merged, even though github is still confused. Thanks!

aaronkavlie-wf commented:
:+1: thanks!

myusuf3 commented:
Any docs on how this is supposed to work! We are looking to not to have to deal with tracking offset anymore.

wvanbergen commented:
Godoc documentation: https://godoc.org/github.com/Shopify/sarama#OffsetManager

This is a building block that stores offsets in Kafka, but doesn't automatically deal with offsets for you. You still have to use this API yourself in your consumer implementation. Example: https://github.com/wvanbergen/kafka/pull/72

myusuf3 commented:
Yeah do using the `OffsetManager` to get where we are at for each topic and feedback into consumer is what I am looking for! Are you guys working toward automatic handling of offset? in that branch you linked?

wvanbergen commented:
Yeah, the idea of that branch is that it deals with offsets for you, but it requires Zookeeper to manage state of the instances. Because of this it's not in sarama itself.

myusuf3 commented:
Awesome! I am going to starting using new offset stuff today! Thanks guys!
